tools/base/aio.py

```diff
@@ -83,3 +105,386 @@ async def run_system_command():
+    yielded in the results.
+
+    If `yield_exceptions` is False (the default), then the wrapped error will be
+    raised immediately.
+
+    If you use any kind of `Generator` or `AsyncGenerator` to produce the
+    awaitables, and `yield_exceptions` is `False`, in the event that an error
+    occurs, it is your responsibility to `close` remaining awaitables that you
+    might have created but which have not already been fired.
+
+    This utility is mostly useful for parallelizing io-bound (as opposed to
+    cpu-bound) tasks.
+
+    Example usage:
+
+    ```
+    import random
+
+    from tools.base import aio
+
+    async def task_to_run(i):
+        print(f"{i} starting")
+        wait = random.random() * 10
+        await asyncio.sleep(wait)
+        return i, wait
+
+    async def run(coros):
+        async for (i, wait) in aio.parallel(coros, limit=3):
+            print(f"{i} waited {wait}")
+
+    def provider():
+        for i in range(0, 10):
+            yield task_to_run(i)
+
+    asyncio.run(run(provider()))
+    ```
+    """
+
+    def __init__(
+            self,
+            coros: Union[types.AsyncGeneratorType, AsyncIterable[Awaitable],
+                         AsyncIterator[Awaitable], types.GeneratorType, Iterator[Awaitable],
+                         Iterable[Awaitable]],
+            yield_exceptions: Optional[bool] = False,
+            limit: Optional[int] = None):
+        self._coros = coros
+        self._limit = limit
+        self._running: List[asyncio.Task] = []
+        self.yield_exceptions = yield_exceptions
+
+    def __aiter__(self) -> AsyncIterator:
+        """Start a coroutine task to process the submit queue, and return
+        an async generator to deliver results back as they arrive
+        """
+        self.submit_task = asyncio.create_task(self.submit())
+        return self.output()
+
+    @property
+    def active(self) -> bool:
+        """Checks whether the iterator is active, either because it
+        hasn't finished submitting or because there are still tasks running
+        """
+        return self.submitting or self.running
+
+    @property
+    def closed(self) -> bool:
+        """If an unhandled error occurs, the generator is closed and no further
+        processing should happen
+        """
+        return self.closing_lock.locked()
+
+    @cached_property
+    def closing_lock(self) -> asyncio.Lock:
+        """Flag to indicate whether the generator has been closed"""
+        return asyncio.Lock()
+
+    @async_property
+    async def coros(self) -> AsyncIterator[Union[AsyncParallelIteratorError, Awaitable]]:
+        """An async iterator of the provided coroutines"""
+        async for coro in self.iter_coros():
+            yield coro
+
+    @property
+    def default_limit(self) -> int:
+        """Default is to use cpu+4 to a max of 32 coroutines"""
+        return min(32, (os.cpu_count() or 0) + 4)
```

Comment:
Out of curiosity, where does this heuristic come from?

Reply:
as explained in the class comment its taken from python's default threadpoolexecutor - completely arbitrary - but possibly not a bad number to use as a default still

ill move the comment down here - so its more obvious

Reply:
apologies - i force of habit pushed adding a comment here